anjani/plugins/spam_prediction.py

```diff
@@ -333,81 +353,44 @@ async def spam_check(self, message: Message, text: str) -> None:
         content_hash = self._build_hash(text)
         identifier = self._build_hex(user)
         proba_str = self.model.prob_to_string(probability)
-        msg = None
+        msg_id = None
 
+        # only log public chat
         if message.chat.username:
-
-            notice = (
-                "#SPAM_PREDICTION\n\n"
-                f"**Prediction Result**: {proba_str}\n"
-                f"**Identifier**: `{identifier}`\n"
+            notice, keyb = await self._build_notice(
+                message, text, proba_str, identifier, content_hash
             )
-            if ch := message.forward_from_chat:
-                notice += f"**Channel ID**: `{self._build_hex(ch.id)}`\n"
 
-            notice += (
-                f"**Message Text Hash**: `{content_hash}`\n\n**====== CONTENT =======**\n\n{text}"
-            )
-            l_spam, l_ham = 0, 0
-            _, data = await asyncio.gather(
-                self.bot.log_stat("predicted"), self.db.find_one({"_id": content_hash})
-            )
-            if data:
-                l_spam = len(data["spam"])
-                l_ham = len(data["ham"])
-
-            keyb = [
-                [
-                    InlineKeyboardButton(
-                        text=f"✅ Correct ({l_spam})", callback_data="spam_check_t"
-                    ),
-                    InlineKeyboardButton(
-                        text=f"❌ Incorrect ({l_ham})", callback_data="spam_check_f"
-                    ),
-                ],
-                [InlineKeyboardButton(text="Chat", url=f"https://t.me/{message.chat.username}")],
-            ]
-
-            if message.forward_from_chat and message.forward_from_chat.username:
-                raw_btn = InlineKeyboardButton(
-                    text="Channel", url=f"https://t.me/{message.forward_from_chat.username}"
-                )
-                if message.chat.username:
-                    keyb[1].append(raw_btn)
-                else:
-                    keyb.append([raw_btn])
+            data = await self.db.find_one({"_id": content_hash})
```

Comment:
i think this need to be lock with `asyncio.Lock` up until the insertion line to avoid race condition and causing insertion more than 1 if spammer goes send messages more than 5 in less than 1 sec